christfellowshipchurch/apollos-app
# [Chat] Moderation

### Summary

This PR adds moderation to the app. There are three types of moderation. After chatting with Stream support some, I think I have a handle on how these are supposed to work.

- **Flag a message.** Flagging a message does nothing to the chat data. Flagged messages appear in the Stream.io dashboard. Dashboard users can then take actions to delete the message, ban the user, delete the user, etc. 
    - Any user can flag a message.
    - There is no flagging of users, 1) because I didn't want to confuse the situation with too many moderation options, and 2) flagged users are supposed to show up in the Dashboard, but they are not presently.


- **Mute a user.** Muting a user is a one-to-one relationship, and shows up in the user data. It is a global setting, so a muted user will be muted forever, across livestream events. Since it's a one-to-one relationship, the message still gets broadcast over the websocket, so the client is supposed to hide it. Right now, I am hiding it at the point of render. I was afraid that filtering it from the array would somehow break pagination. Also, if you unmute a user, all their messages re-appear in place. Frankly, I think Stream should handle muted users, so that the message is gone from the source of truth, but alas. 
    - Any user can mute another user.
    - You can unmute a user.

- **Ban a user.** Banning a user is channel-wide. So, if a moderator bans a user, then that user can no longer send messages to the channel for this livestream. But they will be able to send in the next livestream. Banning app-wide is only possible via the Stream server-side library, so would require a GQL mutation or something. While Stream says that only moderators can ban users, I was able to ban with non-moderator users. I believe this is a Stream bug. In any case, I am enforcing this in the UI by only showing the option to moderators.
    - So only moderators can ban a user.
    - Banning asks for confirmation.
    - There is currently no unban.


### Discussion


| |**My Messages**|**Others' Messages (Non-moderator)**|**Others' Messages (Moderator)**|
|--|--|--|--|
|**Add Reaction**|✅|✅|✅|
|**Edit Message**|✅|||
|**Delete Message**|✅||✅|
|**Flag/Report Message**||✅|✅|
|**(Un)Mute Person**||✅|✅|
|**Ban Person**|||✅|

- Should we have Flag User?
- Do we need Unban User?


### Testing


1. Start two sims. Make sure user A is a moderator and user B is not. You can either make sure that user B is not in the Security Group in Rock, or just hard-code your local API.
1. Send a few messages back-and-forth. When you long press a message, an action sheet shows up. Ensure it adheres to the table above.
    1. Flag a message. It will show up in the Stream.io dashboard (click on the app → Chat → Moderation)
    1. Mute a user. No more messages from that user show up going forward. Unmute user. All hidden messages appear.
    1. Ban a user. That user can no longer send messages. To unban the user—this isn't great—since ban is channel-specific, I've just been deleting the channel via the [Stream CLI](https://getstream.io/chat/docs/cli_introduction/?language=js). 

    ```js
    $ yarn global add getstream-cli
    $ npx stream chat:channel:remove
    # Enter 1e07e4f2e1b52807e43b2338c73d88ba for the channel ID
    ```





### Screenshots

|**My Messages**|**Others' Messages (Non-moderator)**|**Others' Messages (Moderator)**|
|--|--|--|
|<img width="434" alt="Screen Shot 2020-09-08 at 2 39 57 PM" src="https://user-images.githubusercontent.com/200488/92521393-28a91680-f1eb-11ea-9623-4af7156f1d8a.png">|<img width="417" alt="Screen Shot 2020-09-08 at 2 40 16 PM" src="https://user-images.githubusercontent.com/200488/92521405-2c3c9d80-f1eb-11ea-85df-9ec19a2e5516.png">|<img width="434" alt="Screen Shot 2020-09-08 at 2 40 05 PM" src="https://user-images.githubusercontent.com/200488/92521404-2ba40700-f1eb-11ea-8841-ca689b6f556b.png">|
